### PR TITLE
Add CINN Compile Option

### PR DESCRIPTION
**Pull request by zhhsplendid**
### PR types

New features
### PR changes

Others
### Describe

Add CINN compile option in CMake.

Now you can use CINN in Paddle by `-DWITH_CINN=ON` when `cmake`

To test it, you can run `make cinn_lib_test -j` and `ctest -R cinn_lib_test`. 

Note:
1. You should set
```
export runtime_include_dir=${CINN_SOURCE_DIR}/cinn/runtime/cuda 
```
When run test, the `${CINN_SOURCE_DIR}` should be set based on your CINN directory.

2. CINN is under developing now, you may have to change `CINN_GIT_TAG` to the git commit you need.

### CODE REVIEW

File: cmake/cinn.cmake [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  bad_variant_access
+  raw_hash_set
+  )
+set(ABSL_LIB_DIR "${CINN_BINARY_DIR}/dist/third_party/absl/lib")
+set(ABSL_INCLUDE_DIR "${CINN_BINARY_DIR}/dist/third_party/absl/include")
+add_library(absl STATIC IMPORTED GLOBAL)
+set_target_properties(absl PROPERTIES IMPORTED_LOCATION ${ABSL_LIB_DIR}/libabsl_base.a)
+foreach(lib_name ${ABSL_LIB_NAMES})
+    target_link_libraries(absl INTERFACE ${ABSL_LIB_DIR}/libabsl_${lib_name}.a)
+endforeach()
```
Comment (wzzju) [inline cmake/cinn.cmake:77]: Why need these statements?

File: cmake/cinn.cmake [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+foreach(lib_name ${LLVM_LIB_NAMES})
+    target_link_libraries(llvm INTERFACE ${LLVM_LIB_DIR}/libLLVM${lib_name}.a)
+endforeach()
+include_directories(${LLVM_INCLUDE_DIR})
+
+######################################################
+# Put external_cinn and dependencies together as a lib
+######################################################
+
+set(CINN_LIB_NAME "libcinnapi.so")
```
Comment (wzzju) [inline cmake/cinn.cmake:104]: Do we only need the `libcinnapi.so` library? Don’t need the `libcinncore.a`?

File: cmake/cinn.cmake [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+set(LLVM_LIB_NAMES
+  ExecutionEngine
+  )
+set(LLVM_LIB_DIR "${CINN_BINARY_DIR}/dist/third_party/llvm/lib")
+set(LLVM_INCLUDE_DIR "${CINN_BINARY_DIR}/dist/third_party/llvm/include")
+add_library(llvm STATIC IMPORTED GLOBAL)
+set_target_properties(llvm PROPERTIES IMPORTED_LOCATION ${LLVM_LIB_DIR}/libLLVMCore.a)
+foreach(lib_name ${LLVM_LIB_NAMES})
+    target_link_libraries(llvm INTERFACE ${LLVM_LIB_DIR}/libLLVM${lib_name}.a)
+endforeach()
```
Comment (wzzju) [inline cmake/cinn.cmake:97]: 是因为`libLLVMCore.a`中没有`libLLVMExecutionEngine.a`的内容？这里才需要将`libLLVMExecutionEngine.a`的内容打包到`libLLVMCore.a`中吗？

File: cmake/cinn.cmake (hunk 1 above)
Comment (zhhsplendid) [reply cmake/cinn.cmake:77]: cinn动态库需要的依赖

File: cmake/cinn.cmake (hunk 2 above)
Comment (zhhsplendid) [reply cmake/cinn.cmake:104]: 我们只需要libcinnapi.so，但是需要make cinncore，因为头文件是make cinncore时拷贝的

File: cmake/cinn.cmake (hunk 3 above)
Comment (zhhsplendid) [reply cmake/cinn.cmake:97]: 1. 没有。2，但是cinn测试的头文件又使用到了。所以link了libLLVMExecutionEngine.a

**Review by wzzju — APPROVED** [review]
LGTM.

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Pull request merged**